src/__internal__/focus-trap/focus-trap.component.js

```diff
@@ -0,0 +1,118 @@
+import React, {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from "react";
+import PropTypes from "prop-types";
+
+const defaultFocusableSelectors =
+  'button:not([disabled]), [href], input:not([type="hidden"]):not([disabled]), select:not([disabled]), textarea:not([disabled]), [tabindex]';
+
+const FocusTrap = ({
+  children,
+  autoFocus = true,
+  focusFirstElement,
+  bespokeTrap,
+}) => {
+  const ref = useRef();
+  const firstOpen = useRef(true);
+  const [focusableElements, setFocusableElements] = useState();
+  const [firstElement, setFirstElement] = useState();
+  const [lastElement, setLastElement] = useState();
+
+  function setElementFocus(element) {
+    if (typeof element === "function") {
+      element();
+    } else {
+      element.focus();
+    }
+  }
+
+  const hasNewInputs = useCallback(
+    (candidate) => {
+      if (!focusableElements || candidate.length !== focusableElements.length) {
+        return true;
+      }
+
+      return Array.from(candidate).some((el, i) => el !== focusableElements[i]);
+    },
+    [focusableElements]
+  );
+
+  useLayoutEffect(() => {
+    const elements = Array.from(
+      ref.current.querySelectorAll(defaultFocusableSelectors)
+    ).filter((el) => Number(el.tabIndex) !== -1);
+
+    if (hasNewInputs(elements)) {
+      setFocusableElements(Array.from(elements));
+      setFirstElement(elements[0]);
+      setLastElement(elements[elements.length - 1]);
+    }
+  }, [children, hasNewInputs]);
+
+  useEffect(() => {
+    if (autoFocus && firstOpen.current && (focusFirstElement || firstElement)) {
+      setElementFocus(focusFirstElement || firstElement);
+      firstOpen.current = false;
+    }
+  }, [autoFocus, firstElement, focusFirstElement]);
+
+  useEffect(() => {
```

Comment:
Is it necessary to run `ev.preventDefault()` and `setElementFocus` in cases when its not a last/first element?
I think when we're let's say on 3rd input out of 10 in `Dialog`, and press `Tab` then the default browser bevahiour would be just fine wouldnt it?

Reply:
It is necessary to run it unfortunately, as when a disabled element becomes enabled it is not included in the tab order as in the gif in the PR description. Even when you manually set a tabIndex of 0 when it becomes enabled it still doesn't include it until the Modal is closed and re-opened